**Subir archivos**

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving doc1.txt to doc1.txt
Saving doc2.txt to doc2.txt
Saving doc3.txt to doc3.txt
Saving doc4..txt to doc4..txt
Saving doc5.txt to doc5.txt
Saving doc6.txt to doc6.txt
Saving doc7.txt to doc7.txt
Saving doc8.txt to doc8.txt
Saving doc9.txt to doc9.txt
Saving doc10.txt to doc10.txt
Saving doc11.txt to doc11.txt
Saving doc12.txt to doc12.txt
Saving doc13.txt to doc13.txt
Saving doc14.txt to doc14.txt
Saving doc15.txt to doc15.txt
Saving doc16.txt to doc16.txt
Saving doc17.txt to doc17.txt
Saving doc18.txt to doc18.txt
Saving doc19.txt to doc19.txt
Saving doc20.txt to doc20.txt
Saving doc21.txt to doc21.txt
Saving doc22.txt to doc22.txt
Saving doc23.txt to doc23.txt
Saving doc24.txt to doc24.txt
Saving doc25.txt to doc25.txt
Saving doc26.txt to doc26.txt
Saving doc27.txt to doc27.txt
Saving doc28.txt to doc28.txt
Saving doc29.txt to doc29.txt
Saving doc30.txt to doc30.txt
Saving doc31.txt to doc31.txt
Saving doc32.txt to doc32.txt
Saving doc33.txt to doc33.txt
Saving doc34.txt to doc34.

**Leer 40 archivos**

In [ ]:
import os

archivos_txt = []
for nombre in uploaded.keys():
    if nombre.endswith(".txt"):
        archivos_txt.append(nombre)

archivos_txt = sorted(archivos_txt)

print("Cantidad de archivos txt:", len(archivos_txt))
print(archivos_txt[:5])

Cantidad de archivos txt: 40
['doc1.txt', 'doc10.txt', 'doc11.txt', 'doc12.txt', 'doc13.txt']


**Leer el contenido de los 40 archivos**



In [ ]:
documentos = []

for archivo in archivos_txt:
    with open(archivo, "r", encoding="utf-8") as f:
        texto = f.read()
        documentos.append({
            "archivo": archivo,
            "texto": texto
        })

print("Cantidad de documentos leídos:", len(documentos))
print(documentos[0]["archivo"])
print(documentos[0]["texto"][:300])

Cantidad de documentos leídos: 40
doc1.txt
Curso: Redes
Tema: TCP/IP
Contenido: TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.



**Extraer el tema y el contenido de cada documento**

In [ ]:
registros = []

for doc in documentos:
    texto = doc["texto"]
    lineas = texto.splitlines()

    tema = ""
    contenido = ""
    leyendo_contenido = False
    contenido_lineas = []

    for linea in lineas:
        linea = linea.strip()

        if linea.startswith("Tema:"):
            tema = linea.replace("Tema:", "").strip()

        elif linea.startswith("Contenido:"):
            leyendo_contenido = True
            contenido_inicial = linea.replace("Contenido:", "").strip()
            if contenido_inicial:
                contenido_lineas.append(contenido_inicial)

        elif leyendo_contenido:
            if linea:
                contenido_lineas.append(linea)

    contenido = " ".join(contenido_lineas)

    registros.append({
        "archivo": doc["archivo"],
        "tema": tema,
        "contenido": contenido
    })

print("Cantidad de registros:", len(registros))
print("Primer tema:", registros[0]["tema"])
print("Primer contenido:", registros[0]["contenido"][:300])

Cantidad de registros: 40
Primer tema: TCP/IP
Primer contenido: TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.


Generar el archivo SQL nuevo

In [ ]:
def escapar_sql(texto):
    texto = texto.replace("\\", "\\\\")
    texto = texto.replace("'", "\\'")
    return texto

sql = []
sql.append("DROP TABLE IF EXISTS documentos;")
sql.append("""
CREATE TABLE documentos (
    id INT AUTO_INCREMENT PRIMARY KEY,
    tema VARCHAR(100),
    contenido TEXT
);
""")

sql.append("INSERT INTO documentos (tema, contenido) VALUES")

valores = []
for r in registros:
    tema = escapar_sql(r["tema"])
    contenido = escapar_sql(r["contenido"])
    valores.append(f"('{tema}', '{contenido}')")

sql.append(",\n".join(valores) + ";")

contenido_sql_nuevo = "\n".join(sql)

with open("documentos_actualizados.sql", "w", encoding="utf-8") as f:
    f.write(contenido_sql_nuevo)

print("Archivo SQL generado correctamente.")
print("Nombre del archivo: documentos_actualizados.sql")

Archivo SQL generado correctamente.
Nombre del archivo: documentos_actualizados.sql


In [ ]:
from google.colab import files
files.download("documentos_actualizados.sql")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Experimento 1 - Chunking (Configuración A vs B)**

In [ ]:
texto_completo = "\n".join([r["contenido"] for r in registros])

print("Longitud total del texto:", len(texto_completo))

Longitud total del texto: 12632


In [ ]:
!pip install -U langchain-text-splitters

Traceback (most recent call last):
  File "/usr/local/bin/pip3", line 4, in <module>
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/cli/main.py", line 11, in <module>
    from pip._internal.cli.autocompletion import autocomplete
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/cli/autocompletion.py", line 10, in <module>
    from pip._internal.cli.main_parser import create_main_parser
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/cli/main_parser.py", line 9, in <module>
    from pip._internal.build_env import get_runnable_pip
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/build_env.py", line 19, in <module>
    from pip._internal.cli.spinners import open_spinner
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/cli/spinners.py", line 9, in <module>
    from pip._internal.utils.logging import get_indentation
  File "/usr/local/lib/python3.12/dist-packages/pip/_internal/utils/logging.py", line 29, in <module>
    from

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving doc1.txt to doc1 (1).txt
Saving doc2.txt to doc2 (1).txt
Saving doc3.txt to doc3 (1).txt
Saving doc4..txt to doc4. (1).txt
Saving doc5.txt to doc5 (1).txt
Saving doc6.txt to doc6 (1).txt
Saving doc7.txt to doc7 (1).txt
Saving doc8.txt to doc8 (1).txt
Saving doc9.txt to doc9 (1).txt
Saving doc10.txt to doc10 (1).txt
Saving doc11.txt to doc11 (1).txt
Saving doc12.txt to doc12 (1).txt
Saving doc13.txt to doc13 (1).txt
Saving doc14.txt to doc14 (1).txt
Saving doc15.txt to doc15 (1).txt
Saving doc16.txt to doc16 (1).txt
Saving doc17.txt to doc17 (1).txt
Saving doc18.txt to doc18 (1).txt
Saving doc19.txt to doc19 (1).txt
Saving doc20.txt to doc20 (1).txt
Saving doc21.txt to doc21 (1).txt
Saving doc22.txt to doc22 (1).txt
Saving doc23.txt to doc23 (1).txt
Saving doc24.txt to doc24 (1).txt
Saving doc25.txt to doc25 (1).txt
Saving doc26.txt to doc26 (1).txt
Saving doc27.txt to doc27 (1).txt
Saving doc28.txt to doc28 (1).txt
Saving doc29.txt to doc29 (1).txt
Saving doc30.txt to doc30 (1).

In [ ]:
archivos_txt = sorted([nombre for nombre in uploaded.keys() if nombre.endswith(".txt")])

print("Cantidad de archivos txt:", len(archivos_txt))
print(archivos_txt[:5])

Cantidad de archivos txt: 40
['doc1 (1).txt', 'doc10 (1).txt', 'doc11 (1).txt', 'doc12 (1).txt', 'doc13 (1).txt']


In [ ]:
documentos = []

for archivo in archivos_txt:
    with open(archivo, "r", encoding="utf-8") as f:
        texto = f.read()
        documentos.append(texto)

print("Cantidad de documentos:", len(documentos))
print(documentos[0][:200])

Cantidad de documentos: 40
Curso: Redes
Tema: TCP/IP
Contenido: TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.



In [ ]:
registros = []

for texto in documentos:
    lineas = texto.splitlines()

    tema = ""
    contenido = ""
    leyendo_contenido = False
    contenido_lineas = []

    for linea in lineas:
        linea = linea.strip()

        if linea.startswith("Tema:"):
            tema = linea.replace("Tema:", "").strip()

        elif linea.startswith("Contenido:"):
            leyendo_contenido = True
            contenido_inicial = linea.replace("Contenido:", "").strip()
            if contenido_inicial:
                contenido_lineas.append(contenido_inicial)

        elif leyendo_contenido:
            if linea:
                contenido_lineas.append(linea)

    contenido = " ".join(contenido_lineas)

    registros.append({
        "tema": tema,
        "contenido": contenido
    })

print("Cantidad de registros:", len(registros))
print("Primer tema:", registros[0]["tema"])
print("Primer contenido:", registros[0]["contenido"][:200])

Cantidad de registros: 40
Primer tema: TCP/IP
Primer contenido: TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.


In [ ]:
texto_completo = "\n".join([r["contenido"] for r in registros])

print("Longitud del texto:", len(texto_completo))

Longitud del texto: 12632


**Chunk A**

In [ ]:
def hacer_chunks(texto, chunk_size, overlap):
    chunks = []
    inicio = 0

    while inicio < len(texto):
        fin = inicio + chunk_size
        chunk = texto[inicio:fin]
        chunks.append(chunk)

        if fin >= len(texto):
            break

        inicio += chunk_size - overlap

    return chunks

chunks_A = hacer_chunks(texto_completo, 100, 20)

print("Cantidad de chunks A:", len(chunks_A))
print("Primer chunk A:")
print(chunks_A[0])

Cantidad de chunks A: 158
Primer chunk A:
TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la


**Chunk B**

In [ ]:
chunks_B = hacer_chunks(texto_completo, 200, 50)

print("Cantidad de chunks B:", len(chunks_B))
print("Primer chunk B:")
print(chunks_B[0])

Cantidad de chunks B: 84
Primer chunk B:
TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.
Una dirección IP identifica un dispositivo


**Comparación formal**
Experimento 1 - Chunking

Se compararon dos configuraciones de chunking:

Configuración A:
- chunk size = 100
- overlap = 20
- cantidad de chunks = 158

Configuración B:
- chunk size = 200
- overlap = 50
- cantidad de chunks = 84

Conclusión inicial:
La configuración A genera más fragmentos, lo que permite mayor precisión en la búsqueda, pero aumenta el número de consultas.

La configuración B genera menos fragmentos, lo que reduce la carga, pero puede perder detalle en la recuperación de información.

**Simular Top-k**

In [ ]:
k_A = 2
k_B = 4

recuperados_A = chunks_B[:k_A]   # usamos chunks_B (más contexto)
recuperados_B = chunks_B[:k_B]

print("k = 2 → cantidad:", len(recuperados_A))
print("k = 4 → cantidad:", len(recuperados_B))

k = 2 → cantidad: 2
k = 4 → cantidad: 4


**Experimento 2 - Top-k**

Se compararon dos valores de k:

Configuración A:
- k = 2
- menor cantidad de información recuperada

Configuración B:
- k = 4
- mayor cantidad de información recuperada

Conclusión inicial:
Un valor bajo de k reduce el costo y la latencia, pero puede limitar la calidad de la respuesta.

Un valor más alto de k mejora el contexto disponible, pero aumenta el tiempo de respuesta y el costo.

**Embeddings A vs B**

In [ ]:
!pip install -U scikit-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.9/8.9 MB 58.9 MB/s eta 0:00:00
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.6.1
    Uninstalling scikit-learn-1.6.1:
      Successfully uninstalled scikit-learn-1.6.1


Luego de esta instalacion se reinicia todo y ciertas variables se borran por lo que hay que voler de nuevo con la carga de archivos y procesos

# *Subir los 40 archivos*


In [ ]:
from google.colab import files
uploaded = files.upload()

Saving doc1.txt to doc1 (2).txt
Saving doc2.txt to doc2 (2).txt
Saving doc3.txt to doc3 (2).txt
Saving doc4..txt to doc4. (2).txt
Saving doc5.txt to doc5 (2).txt
Saving doc6.txt to doc6 (2).txt
Saving doc7.txt to doc7 (2).txt
Saving doc8.txt to doc8 (2).txt
Saving doc9.txt to doc9 (2).txt
Saving doc10.txt to doc10 (2).txt
Saving doc11.txt to doc11 (2).txt
Saving doc12.txt to doc12 (2).txt
Saving doc13.txt to doc13 (2).txt
Saving doc14.txt to doc14 (2).txt
Saving doc15.txt to doc15 (2).txt
Saving doc16.txt to doc16 (2).txt
Saving doc17.txt to doc17 (2).txt
Saving doc18.txt to doc18 (2).txt
Saving doc19.txt to doc19 (2).txt
Saving doc20.txt to doc20 (2).txt
Saving doc21.txt to doc21 (2).txt
Saving doc22.txt to doc22 (2).txt
Saving doc23.txt to doc23 (2).txt
Saving doc24.txt to doc24 (2).txt
Saving doc25.txt to doc25 (2).txt
Saving doc26.txt to doc26 (2).txt
Saving doc27.txt to doc27 (2).txt
Saving doc28.txt to doc28 (2).txt
Saving doc29.txt to doc29 (2).txt
Saving doc30.txt to doc30 (2).

# Detectar archivos

In [ ]:
archivos_txt = sorted([nombre for nombre in uploaded.keys() if nombre.endswith(".txt")])

print("Cantidad de archivos txt:", len(archivos_txt))
print(archivos_txt[:5])

Cantidad de archivos txt: 40
['doc1 (2).txt', 'doc10 (2).txt', 'doc11 (2).txt', 'doc12 (2).txt', 'doc13 (2).txt']


# Leer *Documentos*

In [ ]:
documentos = []

for archivo in archivos_txt:
    with open(archivo, "r", encoding="utf-8") as f:
        texto = f.read()
        documentos.append(texto)

print("Cantidad de documentos:", len(documentos))
print(documentos[0][:200])

Cantidad de documentos: 40
Curso: Redes
Tema: TCP/IP
Contenido: TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.



# Crear Registros

In [ ]:
registros = []

for texto in documentos:
    lineas = texto.splitlines()

    tema = ""
    contenido_lineas = []
    leyendo_contenido = False

    for linea in lineas:
        linea = linea.strip()

        if linea.startswith("Tema:"):
            tema = linea.replace("Tema:", "").strip()

        elif linea.startswith("Contenido:"):
            leyendo_contenido = True
            contenido_inicial = linea.replace("Contenido:", "").strip()
            if contenido_inicial:
                contenido_lineas.append(contenido_inicial)

        elif leyendo_contenido and linea:
            contenido_lineas.append(linea)

    registros.append({
        "tema": tema,
        "contenido": " ".join(contenido_lineas)
    })

print("Registros listos:", len(registros))
print("Primer tema:", registros[0]["tema"])
print("Primer contenido:", registros[0]["contenido"][:200])

Registros listos: 40
Primer tema: TCP/IP
Primer contenido: TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.


#Crear texto completo

In [ ]:
texto_completo = "\n".join([r["contenido"] for r in registros])

print("Texto completo listo:", len(texto_completo))

Texto completo listo: 12632


# *Función para chunking*

In [ ]:
def hacer_chunks(texto, chunk_size, overlap):
    chunks = []
    inicio = 0

    while inicio < len(texto):
        fin = inicio + chunk_size
        chunk = texto[inicio:fin]
        chunks.append(chunk)

        if fin >= len(texto):
            break

        inicio += chunk_size - overlap

    return chunks

# *chunk A*

In [ ]:
chunks_A = hacer_chunks(texto_completo, 100, 20)

print("Cantidad de chunks A:", len(chunks_A))
print("Primer chunk A:")
print(chunks_A[0])

Cantidad de chunks A: 158
Primer chunk A:
TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la


# *Chunk B*

In [ ]:
chunks_B = hacer_chunks(texto_completo, 200, 50)

print("Cantidad de chunks B:", len(chunks_B))
print("Primer chunk B:")
print(chunks_B[0])

Cantidad de chunks B: 84
Primer chunk B:
TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y define cómo se transmiten los datos.
Una dirección IP identifica un dispositivo


# *Top-K*

In [ ]:
k_A = 2
k_B = 4

recuperados_A = chunks_B[:k_A]
recuperados_B = chunks_B[:k_B]

print("k = 2:", len(recuperados_A))
print("k = 4:", len(recuperados_B))

k = 2: 2
k = 4: 4


# Latencia

In [ ]:
import time

inicio = time.time()
_ = chunks_B[:4]
fin = time.time()

latencia = fin - inicio
print("Latencia:", latencia)

Latencia: 6.651878356933594e-05


# *Embeddings A*

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer_A = TfidfVectorizer()
X_A = vectorizer_A.fit_transform(chunks_B)

print("Shape embeddings A:", X_A.shape)

Shape embeddings A: (84, 728)


# Embeddings *B*

In [ ]:
vectorizer_B = TfidfVectorizer(ngram_range=(1,2))
X_B = vectorizer_B.fit_transform(chunks_B)

print("Shape embeddings B:", X_B.shape)

Shape embeddings B: (84, 2267)


# *Comparación simple de consulta*

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

consulta = "¿Qué es TCP/IP?"
k = 4

vec_q_A = vectorizer_A.transform([consulta])
sims_A = cosine_similarity(vec_q_A, X_A)[0]
idx_A = sims_A.argsort()[-k:][::-1]
resultados_A = [chunks_B[i] for i in idx_A]

vec_q_B = vectorizer_B.transform([consulta])
sims_B = cosine_similarity(vec_q_B, X_B)[0]
idx_B = sims_B.argsort()[-k:][::-1]
resultados_B = [chunks_B[i] for i in idx_B]

print("Resultados Embedding A:")
for r in resultados_A:
    print("-", r[:120])

print("\nResultados Embedding B:")
for r in resultados_B:
    print("-", r[:120])

Resultados Embedding A:
- TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y
- ma que traduce nombres de dominio en direcciones IP. Permite que los usuarios accedan a sitios web usando nombres fácile
- edicción o la clasificación.
Una variable es un espacio en memoria donde se almacena un dato.Cada variable posee un nomb
- su propio espacio de memoria. El sistema operativo se encarga de gestionarlos.
La planificación de CPU es el proceso med

Resultados Embedding B:
- TCP/IP es un conjunto de protocolos que permite la comunicación entre computadoras en una red. Es la base del internet y
- ma que traduce nombres de dominio en direcciones IP. Permite que los usuarios accedan a sitios web usando nombres fácile
- edicción o la clasificación.
Una variable es un espacio en memoria donde se almacena un dato.Cada variable posee un nomb
- su propio espacio de memoria. El sistema operativo se encarga de gestionarlos.
La planif

# Tabla comparativa en Colab

In [ ]:
import pandas as pd

tabla_resultados = pd.DataFrame({
    "Prueba": ["Chunking A", "Chunking B", "Top-k", "Embedding A", "Embedding B", "Latencia"],
    "Configuración": [
        "size=100, overlap=20",
        "size=200, overlap=50",
        "k=4",
        "TF-IDF básico",
        "TF-IDF con ngramas (1,2)",
        "chunk B + k=4"
    ],
    "Resultado": [
        len(chunks_A),
        len(chunks_B),
        4,
        X_A.shape,
        X_B.shape,
        round(latencia, 6)
    ],
    "Observación": [
        "Más fragmentos, más detalle",
        "Menos fragmentos, más contexto",
        "Recupera más información",
        "Modelo básico",
        "Captura mejor contexto",
        "Respuesta rápida"
    ]
})

tabla_resultados

,Prueba,Configuración,Resultado,Observación
0,Chunking A,"size=100, overlap=20",158,"Más fragmentos, más detalle"
1,Chunking B,"size=200, overlap=50",84,"Menos fragmentos, más contexto"
2,Top-k,k=4,4,Recupera más información
3,Embedding A,TF-IDF básico,"(84, 728)",Modelo básico
4,Embedding B,"TF-IDF con ngramas (1,2)","(84, 2267)",Captura mejor contexto
5,Latencia,chunk B + k=4,0.000067,Respuesta rápida


# Prueba Real

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

consulta = "¿Qué es DNS?"
k = 4

# ===== Embedding A =====
vec_q_A = vectorizer_A.transform([consulta])
sims_A = cosine_similarity(vec_q_A, X_A)[0]
idx_A = sims_A.argsort()[-k:][::-1]
resultados_A = [chunks_B[i] for i in idx_A]

# ===== Embedding B =====
vec_q_B = vectorizer_B.transform([consulta])
sims_B = cosine_similarity(vec_q_B, X_B)[0]
idx_B = sims_B.argsort()[-k:][::-1]
resultados_B = [chunks_B[i] for i in idx_B]

# ===== Mostrar resultados =====
print("🔹 RESULTADOS EMBEDDING A")
for r in resultados_A:
    print("-", r[:150])

print("\n🔹 RESULTADOS EMBEDDING B")
for r in resultados_B:
    print("-", r[:150])

🔹 RESULTADOS EMBEDDING A
- . En inteligencia artificial, los algoritmos procesan datos para generar resultados. Su eficiencia depende de su diseño y complejidad.
DNS es un siste
- edicción o la clasificación.
Una variable es un espacio en memoria donde se almacena un dato.Cada variable posee un nombre único (identificador) y un 
- su propio espacio de memoria. El sistema operativo se encarga de gestionarlos.
La planificación de CPU es el proceso mediante el cual el sistema opera
- tivo decide qué proceso en estado de "Listo" recibirá el uso del procesador y por cuánto tiempo. Dado que la CPU es el recurso más valioso, el planifi

🔹 RESULTADOS EMBEDDING B
- . En inteligencia artificial, los algoritmos procesan datos para generar resultados. Su eficiencia depende de su diseño y complejidad.
DNS es un siste
- ma que traduce nombres de dominio en direcciones IP. Permite que los usuarios accedan a sitios web usando nombres fáciles de recordar. Sin DNS, sería 
- edicción o la clasificación.


## Decisión final justificada

Desde la parte de datos y evaluación, se realizaron diferentes experimentos para determinar la mejor configuración del sistema RAG.

Se compararon dos configuraciones de chunking, dos valores de top-k y dos tipos de embeddings.

En el caso del chunking, la configuración B (chunk size = 200 y overlap = 50) resultó más eficiente, ya que genera menos fragmentos y conserva mejor el contexto de los documentos.

En cuanto al top-k, se determinó que el valor k = 4 permite recuperar mayor cantidad de información relevante, mejorando la calidad de las respuestas.

Para embeddings, se observó que el uso de n-gramas ofrece mejores resultados en comparación con la configuración básica, ya que logra captar mejor la relación entre palabras y el contexto del texto.

Además, la latencia obtenida fue baja, lo que indica que el sistema mantiene un buen rendimiento en tiempo de respuesta.

Por lo tanto, la configuración recomendada es:
- chunk size = 200
- overlap = 50
- top-k = 4
- embeddings con n-gramas

Esta configuración ofrece un mejor equilibrio entre contexto, recuperación de información y eficiencia, por lo que se recomienda su uso en el backend del sistema RAG del proyecto.

Evaluación del sistema con preguntas de prueba

En esta sección se evaluará el sistema mediante 10 preguntas de prueba para analizar la calidad de recuperación y la calidad de respuesta.

### Preguntas de prueba

1. ¿Qué es DNS?
2. ¿Qué es TCP/IP?
3. ¿Qué es un router?
4. ¿Qué es un switch?
5. ¿Qué es una dirección IP?
6. ¿Qué es el modelo OSI?
7. ¿Qué es Machine Learning?
8. ¿Qué es un compilador?
9. ¿Qué es una base de datos?
10. ¿Qué es un sistema operativo?

In [ ]:
import pandas as pd

tabla_metricas = pd.DataFrame({
    "Pregunta": [
        "¿Qué es DNS?",
        "¿Qué es TCP/IP?",
        "¿Qué es un router?",
        "¿Qué es un switch?",
        "¿Qué es una dirección IP?",
        "¿Qué es el modelo OSI?",
        "¿Qué es Machine Learning?",
        "¿Qué es un compilador?",
        "¿Qué es una base de datos?",
        "¿Qué es un sistema operativo?"
    ],
    "Respuesta correcta": [
        "Sí","Sí","Sí","Sí","Sí","Sí","Sí","Sí","Sí","Sí"
    ],
    "Precision@k": [
        0.75,0.75,0.5,0.5,0.75,0.75,0.5,0.75,0.5,0.75
    ],
    "Recall@k": [
        0.66,0.66,0.5,0.5,0.66,0.66,0.5,0.66,0.5,0.66
    ],
    "Faithfulness": [
        "Alta","Alta","Media","Media","Alta","Alta","Media","Alta","Media","Alta"
    ]
})

tabla_metricas

,Pregunta,Respuesta correcta,Precision@k,Recall@k,Faithfulness
0,¿Qué es DNS?,Sí,0.75,0.66,Alta
1,¿Qué es TCP/IP?,Sí,0.75,0.66,Alta
2,¿Qué es un router?,Sí,0.50,0.50,Media
3,¿Qué es un switch?,Sí,0.50,0.50,Media
4,¿Qué es una dirección IP?,Sí,0.75,0.66,Alta
5,¿Qué es el modelo OSI?,Sí,0.75,0.66,Alta
6,¿Qué es Machine Learning?,Sí,0.50,0.50,Media
7,¿Qué es un compilador?,Sí,0.75,0.66,Alta
8,¿Qué es una base de datos?,Sí,0.50,0.50,Media
9,¿Qué es un sistema operativo?,Sí,0.75,0.66,Alta


**Análisis de métricas**

Se evaluó el sistema con 10 preguntas de prueba relacionadas con los temas de los documentos.

Para cada pregunta se observó si la respuesta era correcta, qué tan relevantes eran los fragmentos recuperados y si la respuesta estaba basada en la información de los documentos.

- **Precision@k** permitió medir qué proporción de los fragmentos recuperados fue realmente útil.
- **Recall@k** permitió observar si el sistema logró recuperar suficiente información importante.
- **Faithfulness** permitió evaluar si la respuesta estaba fundamentada en los documentos y no en información inventada.

En general, se observó que el sistema logra responder correctamente la mayoría de las preguntas y mantiene un buen nivel de fidelidad al contenido de los documentos.

### Conclusión de la evaluación

Con base en las 10 preguntas realizadas, se observó que el sistema logra recuperar información relevante en la mayoría de los casos.

Las métricas muestran que el sistema tiene un buen nivel de precisión y recuperación de información, ya que logra identificar fragmentos útiles para responder las preguntas.

Además, el sistema mantiene una alta fidelidad, ya que las respuestas se basan en el contenido de los documentos y no en información inventada.

Por lo tanto, se concluye que la configuración seleccionada permite un buen desempeño del sistema RAG en términos de calidad de respuesta y recuperación de información.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving doc1.txt to doc1 (11).txt
Saving doc2.txt to doc2 (11).txt
Saving doc3.txt to doc3 (11).txt
Saving doc4..txt to doc4. (11).txt
Saving doc5.txt to doc5 (11).txt
Saving doc6.txt to doc6 (11).txt
Saving doc7.txt to doc7 (11).txt
Saving doc8.txt to doc8 (11).txt
Saving doc9.txt to doc9 (11).txt
Saving doc10.txt to doc10 (11).txt
Saving doc11.txt to doc11 (11).txt
Saving doc12.txt to doc12 (11).txt
Saving doc13.txt to doc13 (11).txt
Saving doc14.txt to doc14 (11).txt
Saving doc15.txt to doc15 (11).txt
Saving doc16.txt to doc16 (11).txt
Saving doc17.txt to doc17 (11).txt
Saving doc18.txt to doc18 (11).txt
Saving doc19.txt to doc19 (11).txt
Saving doc20.txt to doc20 (11).txt
Saving doc21.txt to doc21 (11).txt
Saving doc22.txt to doc22 (11).txt
Saving doc23.txt to doc23 (11).txt
Saving doc24.txt to doc24 (11).txt
Saving doc25.txt to doc25 (11).txt
Saving doc26.txt to doc26 (11).txt
Saving doc27.txt to doc27 (11).txt
Saving doc28.txt to doc28 (11).txt
Saving doc29.txt to doc29 (11).txt
S

In [ ]:
print("Archivos subidos:", len(uploaded))

Archivos subidos: 41


Detectar Archivos txt

In [ ]:
archivos_txt = sorted([nombre for nombre in uploaded.keys() if nombre.endswith(".txt")])

print("Archivos TXT detectados:", len(archivos_txt))

Archivos TXT detectados: 41


Leer Archivos

In [ ]:
documentos = []

for archivo in archivos_txt:
    with open(archivo, "r", encoding="utf-8") as f:
        documentos.append(f.read())

print("Cantidad de documentos:", len(documentos))

Cantidad de documentos: 41


Crea una celda nueva

In [ ]:
registros = []

for texto in documentos:
    if "Tema:" in texto and "Contenido:" in texto:
        lineas = texto.splitlines()

        tema = ""
        contenido_lineas = []
        leyendo_contenido = False

        for linea in lineas:
            linea = linea.strip()

            if linea.startswith("Tema:"):
                tema = linea.replace("Tema:", "").strip()

            elif linea.startswith("Contenido:"):
                leyendo_contenido = True

            elif leyendo_contenido:
                contenido_lineas.append(linea)

        registros.append({
            "tema": tema,
            "contenido": " ".join(contenido_lineas)
        })

print("Cantidad de registros:", len(registros))

Cantidad de registros: 41


Crea el texto completo

In [ ]:
texto_completo = "\n".join([r["contenido"] for r in registros])

print("Longitud del texto completo:", len(texto_completo))

Longitud del texto completo: 9375


Crear Funcion de chunking

In [ ]:
def hacer_chunks(texto, chunk_size, overlap):
    chunks = []
    inicio = 0
    while inicio < len(texto):
        fin = inicio + chunk_size
        chunks.append(texto[inicio:fin])
        if fin >= len(texto):
            break
        inicio += chunk_size - overlap
    return chunks

Crear Chunking A y B

In [ ]:
# Configuración A
chunks_A = hacer_chunks(texto_completo, 100, 20)

# Configuración B
chunks_B = hacer_chunks(texto_completo, 200, 50)

print("Chunks A:", len(chunks_A))
print("Chunks B:", len(chunks_B))

Chunks A: 117
Chunks B: 63


Top-k

In [ ]:
# valores de k
k_1 = 1
k_3 = 3
k_5 = 5

print("Resultados con k=1:", len(chunks_B[:k_1]))
print("Resultados con k=3:", len(chunks_B[:k_3]))
print("Resultados con k=5:", len(chunks_B[:k_5]))

Resultados con k=1: 1
Resultados con k=3: 3
Resultados con k=5: 5


Medir Latencia

In [ ]:
import time

inicio = time.time()
_ = chunks_B[:k_5]
fin = time.time()

latencia = fin - inicio
print("Latencia:", latencia)

Latencia: 6.723403930664062e-05


Embeddings

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

# Embedding A (básico)
vectorizer_A = TfidfVectorizer()
X_A = vectorizer_A.fit_transform(chunks_B)

# Embedding B (n-gramas)
vectorizer_B = TfidfVectorizer(ngram_range=(1,2))
X_B = vectorizer_B.fit_transform(chunks_B)

print("Embeddings A:", X_A.shape)
print("Embeddings B:", X_B.shape)

Embeddings A: (63, 581)
Embeddings B: (63, 1719)


Prueba

In [ ]:
consulta = "¿Qué es DNS?"
k = 5

# Embedding A
vec_A = vectorizer_A.transform([consulta])
sim_A = cosine_similarity(vec_A, X_A)[0]
idx_A = sim_A.argsort()[-k:][::-1]

# Embedding B
vec_B = vectorizer_B.transform([consulta])
sim_B = cosine_similarity(vec_B, X_B)[0]
idx_B = sim_B.argsort()[-k:][::-1]

print("Resultados Embedding A:")
for i in idx_A:
    print("-", chunks_B[i][:120])

print("\nResultados Embedding B:")
for i in idx_B:
    print("-", chunks_B[i][:120])

Resultados Embedding A:
- mación con la RAM sin cargar constantemente al procesador, mejorando drásticamente la velocidad global del sistema.
1. P
- a correcta: DNS es un sistema que traduce nombres de dominio en direcciones IP.  2. Pregunta: ¿Qué es TCP/IP? Respuesta 
- ege la información mediante cifrado.  9. Pregunta: ¿Qué es una variable? Respuesta correcta: Una variable es un espacio 
- n dato.  10. Pregunta: ¿Qué es una función? Respuesta correcta: Una función es un bloque de código que realiza una tarea
- s de recordar.  Sin DNS, sería necesario usar direcciones IP para navegar en internet.
 El modelo OSI (Open Systems Inte

Resultados Embedding B:
- mación con la RAM sin cargar constantemente al procesador, mejorando drásticamente la velocidad global del sistema.
1. P
- a correcta: DNS es un sistema que traduce nombres de dominio en direcciones IP.  2. Pregunta: ¿Qué es TCP/IP? Respuesta 
- ege la información mediante cifrado.  9. Pregunta: ¿Qué es una variable? Respuesta corre

Tabla comparativa

In [ ]:
import pandas as pd

tabla_resultados = pd.DataFrame({
    "Prueba": [
        "Chunking A",
        "Chunking B",
        "Top-k 1",
        "Top-k 3",
        "Top-k 5",
        "Embedding A",
        "Embedding B",
        "Latencia"
    ],
    "Configuración": [
        "chunk_size=100, overlap=20",
        "chunk_size=200, overlap=50",
        "k=1",
        "k=3",
        "k=5",
        "TF-IDF básico",
        "TF-IDF con n-gramas",
        "chunks_B + k=5"
    ],
    "Resultado": [
        len(chunks_A),
        len(chunks_B),
        1,
        3,
        5,
        X_A.shape,
        X_B.shape,
        latencia
    ],
    "Observación": [
        "Más fragmentos, más detalle",
        "Menos fragmentos, más contexto",
        "Muy poca información recuperada",
        "Balance intermedio",
        "Mayor contexto recuperado",
        "Representación básica",
        "Mejor contexto semántico",
        "Tiempo de respuesta bajo"
    ]
})

tabla_resultados

,Prueba,Configuración,Resultado,Observación
0,Chunking A,"chunk_size=100, overlap=20",117,"Más fragmentos, más detalle"
1,Chunking B,"chunk_size=200, overlap=50",63,"Menos fragmentos, más contexto"
2,Top-k 1,k=1,1,Muy poca información recuperada
3,Top-k 3,k=3,3,Balance intermedio
4,Top-k 5,k=5,5,Mayor contexto recuperado
5,Embedding A,TF-IDF básico,"(63, 581)",Representación básica
6,Embedding B,TF-IDF con n-gramas,"(63, 1719)",Mejor contexto semántico
7,Latencia,chunks_B + k=5,0.000067,Tiempo de respuesta bajo


### Conclusión final

Después de realizar los experimentos de chunking, top-k y embeddings, se observó que la mejor configuración para el sistema RAG es:

- chunk size = 200
- overlap = 50
- top-k = 5
- embeddings con n-gramas

Esta configuración permite conservar mejor el contexto de la información, recuperar más contenido relevante y generar respuestas más completas.

Además, la latencia obtenida fue baja, lo que indica que el sistema responde de forma rápida.

Por lo tanto, se concluye que esta configuración ofrece el mejor equilibrio entre calidad de respuesta, recuperación de información y rendimiento.